### PR TITLE
Add support for Wasm C API

### PR DESCRIPTION
**Pull request by MarkMcCaskey**
Begin adding support for the standard Wasm C API

# Review

- [x] Add a short description of the the change to the CHANGELOG.md file

### CODE REVIEW

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try

Expecting test failure, if so, then we're good to go once I edit out the intentional failure

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try-

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try-

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try-

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try

File: lib/runtime-c-api/crates/wasm-c-api/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,1355 @@
...
+use wasmer::compiler::compile;
+use wasmer::error::{CallError, RuntimeError};
+use wasmer::import::{ImportObject, LikeNamespace, Namespace};
+use wasmer::module::Module;
+use wasmer::units;
+use wasmer::wasm;
+
+// TODO: remove delete from macro generation, need to do that manually
+
+// TODO: investigate marking things like the C++ API does (does not work in return position)
```
Comment (Hywan) [inline lib/runtime-c-api/crates/wasm-c-api/src/lib.rs:20]: What do you mean?

File: lib/runtime-c-api/crates/wasm-c-api/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,1355 @@
...
+#[no_mangle]
+pub extern "C" fn wasm_engine_new_with_config(
+    _config_ptr: *mut wasm_config_t,
+) -> *mut wasm_engine_t {
+    wasm_engine_new()
+}
+
+#[repr(C)]
+pub struct wasm_instance_t {
+    real_instance: Arc<wasm::Instance>,
```
Comment (Hywan) [inline lib/runtime-c-api/crates/wasm-c-api/src/lib.rs:74]:
```diff
- real_instance
+ inner
```

?

File: lib/runtime-c-api/crates/wasm-c-api/src/lib.rs (hunk 1 above)
Comment (MarkMcCaskey) [reply lib/runtime-c-api/crates/wasm-c-api/src/lib.rs:20]: In the C++ API they use `own<T>` to represent the transfer of ownership ; I wasn't able to get this to work in the return position using `#[repr(transparent)]` in Rust

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by syrusakbary** [comment]
The `wasm-c-api` is now fully supported in Wasmer `master` thanks to latest refactor. Closing PR!

**Pull request closed without merging**